### PR TITLE
feat(zk): add committing permutations

### PR DESCRIPTION
**Pull request by dongchangYoo**
This PR implements `Commit()` function of `PermutationArgument` and its output structure
See [PermutationArgument::Commit](https://github.com/kroma-network/halo2/blob/7d0a36990452c8e7ebd600de258420781a9b7917/halo2_proofs/src/plonk/permutation/prover.rs#L43-L188), and [PermutationCommitted](https://github.com/kroma-network/halo2/blob/7d0a36990452c8e7ebd600de258420781a9b7917/halo2_proofs/src/plonk/permutation/prover.rs#L21-L28) and its [Construct](https://github.com/kroma-network/halo2/blob/7d0a36990452c8e7ebd600de258420781a9b7917/halo2_proofs/src/plonk/permutation/prover.rs#L190-L203) method.

### CODE REVIEW

File: tachyon/zk/plonk/permutation/permutation_argument.h [hunk 1]
```diff
@@ -16,6 +16,7 @@
 namespace tachyon::zk {
 
 // TODO(dongchangYoo): impl logics related to prove and verify.
+template <typename PCSTy>
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/permutation_argument.h:25]: Why this use `PCSTy` here in this commit?

**Comment by chokobole** [comment]
"test(zk): make the tests in zk/permutation use Halo2ProverTest" commit has 2 changes
- one is "add template type to PermutationArgument"
- the other is "use Halo2ProverTest".

File: tachyon/zk/plonk/permutation/lookup_table.h [hunk 2]
```diff
@@ -56,21 +55,31 @@ class LookupTable {
 
     // Assign [𝛿ⁱw⁰, 𝛿ⁱw¹, 𝛿ⁱw², ..., 𝛿ⁱwⁿ⁻¹] to each row.
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/lookup_table.h:56]: You need to update comment together.

File: tachyon/zk/plonk/permutation/lookup_table.h [hunk 3]
```diff
@@ -10,13 +10,11 @@
...
 #include <vector>
 
-#include "gtest/gtest_prod.h"
-
 #include "tachyon/base/containers/container_util.h"
+#include "tachyon/zk/base/ref.h"
 #include "tachyon/zk/plonk/permutation/label.h"
 
 namespace tachyon::zk {
-
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/lookup_table.h:19]: Why do you remove this blank?

**Comment by chokobole** [comment]
Please remove "refac(zk): convert type from ColumnData to AnyColmun" commit from list. Those functions are intended to use it as `ColumnData`.

File: tachyon/zk/plonk/permutation/column_store.h [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  using ColumnTy = std::vector<F>;
+
+  ColumnStore() = default;
+  ColumnStore(const Columns<Evals>& columns,
+              const std::vector<AnyColumn>& colum_metas,
+              const std::vector<Evals>& permutations, size_t chunk_size,
+              const LookupTable<PCSTy>& original_table)
+      : columns_(columns),
+        column_metas_(absl::MakeSpan(colum_metas)),
+        permutations_(absl::MakeSpan(permutations)),
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/column_store.h:39]:
```suggestion
        column_metas_(absl::MakeConstSpan(colum_metas)),
        permutations_(absl::MakeConstSpan(permutations)),
```

File: tachyon/zk/plonk/permutation/column_store.h [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+template <typename PCSTy>
+class ColumnStore {
+ public:
+  using F = typename PCSTy::Field;
+  using Evals = typename PCSTy::Evals;
+  using ColumnTy = std::vector<F>;
+
+  ColumnStore() = default;
+  ColumnStore(const Columns<Evals>& columns,
+              const std::vector<AnyColumn>& colum_metas,
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/column_store.h:34]:
```suggestion
              const std::vector<AnyColumn>& column_metas,
```

File: tachyon/zk/plonk/permutation/column_store_unittest.cc [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    columns_meta_ = {FixedColumn(0),  InstanceColumn(0), AdviceColumn(0),
+                     AdviceColumn(1), FixedColumn(1),    FixedColumn(2),
+                     AdviceColumn(2), InstanceColumn(1), InstanceColumn(2)};
+
+    // Prepare columns sorted in the order of |columns_meta_|
+    expected_columns_ = {
+        fixed_columns_[0],  instance_columns_[0], advice_columns_[0],
+        advice_columns_[1], fixed_columns_[1],    fixed_columns_[2],
+        advice_columns_[2], instance_columns_[1], instance_columns_[2],
+    };
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/column_store_unittest.cc:46]: Please move them down. Here only contains code shared by tests.

File: tachyon/zk/plonk/permutation/column_store.h [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  size_t chunk_size() const { return chunk_size_; }
+
+  size_t GetChunkNum() const {
+    return (column_metas_.size() + chunk_size_ - 1) / chunk_size_;
+  }
+
+  size_t GetEffectiveChunkSize(size_t chunk_idx) const {
+    size_t total_size = permutations_.size();
+    size_t chunk_num = GetChunkNum();
+    CHECK_LE(chunk_idx, chunk_num);
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/column_store.h:52]:
```suggestion
    CHECK_LT(chunk_idx, chunk_num);
```

File: tachyon/zk/plonk/permutation/column_store.h [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+namespace tachyon::zk {
+
+// |ColumnStore| has three tables with the same number of columns and rows
+// exist. In reality, for optimization purposes, it only has pointers to
+// tables. And |ColumnStore| supports functions that can query the columns of
+// each table coressponding specific chunk.
+// - |original|: 2-dimensinal vector with δ and w labels.
+// - |permuted|: permuted |original|
+// - |columns|: witness columns.
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/column_store.h:21]: Please update the comments.

File: tachyon/zk/plonk/permutation/column_store.h [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    return ret;
+  }
+
+  std::vector<Ref<const ColumnTy>> GetOriginalChunk(size_t chunk_idx) const {
+    const std::vector<ColumnTy>& original_table = original_table_.table();
+
+    size_t start = chunk_idx * chunk_size_;
+    size_t effective_chunk_size = GetEffectiveChunkSize(chunk_idx);
+
+    std::vector<Ref<const ColumnTy>> ret;
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/column_store.h:90]:
```suggestion
    std::vector<Ref<const ColumnTy>> ret;
    ret.reserve(effective_chunk_size);
```

File: tachyon/zk/plonk/permutation/column_store.h [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+    std::vector<Ref<const Evals>> ret;
+    ret.reserve(effective_chunk_size);
+    for (size_t i = 0; i < effective_chunk_size; ++i) {
+      ret.push_back(Ref<const Evals>(&permutations_[start + i]));
+    }
+    return ret;
+  }
+
+  std::vector<Ref<const ColumnTy>> GetOriginalChunk(size_t chunk_idx) const {
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/column_store.h:84]: You can make a shared code between `GetPermutationChunk` and `GetOriginalChunk`

File: tachyon/zk/plonk/permutation/permutation_committed.h [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+class PermutationCommittedSet {
+ public:
+  using F = typename PCSTy::Field;
+  using Poly = typename PCSTy::Poly;
+
+  explicit PermutationCommittedSet(BlindedPolynomial<Poly>& blinded_poly)
+      : blinded_zp_poly_(blinded_poly) {}
+
+  const Poly& zp_poly() const { return blinded_zp_poly_.poly(); }
+  const F& zp_blind() const { return blinded_zp_poly_.blind(); }
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/permutation_committed.h:21]: Do we need this? Plus, snake_case is only allowed for member getter or setter.

File: tachyon/zk/plonk/permutation/permutation_committed.h [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+  const BlindedPolynomial<Poly> blinded_zp_poly() const {
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/permutation_committed.h:22]:
```suggestion
  const BlindedPolynomial<Poly>& blinded_zp_poly() const {
```

File: tachyon/zk/plonk/permutation/permutation_committed.h [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+  BlindedPolynomial<Poly> blinded_zp_poly_;
+};
+
+// A committed polynomials for chunks of columns.
+template <typename PCSTy>
+class PermutationCommitted {
+ public:
+  PermutationCommitted(
+      std::vector<PermutationCommittedSet<PCSTy>> committed_sets)
+      : committed_sets_(committed_sets) {}
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/permutation_committed.h:36]:
```suggestion
      : committed_sets_(std::move(committed_sets)) {}
```

File: tachyon/zk/plonk/permutation/permutation_argument.h [hunk 14]
```diff
@@ -69,8 +88,139 @@ class PermutationArgument {
     return 3;
   }
 
+  // Returns commitments of Zₚ,ᵢ for chunk index i.
+  //
+  // See Halo2 book to figure out logic in detail.
+  // https://zcash.github.io/halo2/design/proving-system/permutation.html
+  PermutationCommitted<PCSTy> Commit(
+      Halo2Prover<PCSTy>* prover, Columns<Evals>& columns,
```
Comment (chokobole) [inline tachyon/zk/plonk/permutation/permutation_argument.h:96]: Please do not use `Halo2Prover` directly

**Review by chokobole — CHANGES_REQUESTED** [review]

**Comment by chokobole** [comment]
Change commit scope `feat(test): use Halo2ProverTest in tests related to permutation` to `test(zk): ...`

**Review by chokobole — APPROVED** [review]
LGTM

File: tachyon/zk/plonk/permutation/column_store.h (hunk 8 above)
Comment (dongchangYoo) [reply tachyon/zk/plonk/permutation/column_store.h]: I think it is resolved. Plz close it @chokobole

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Pull request merged**